File: Code Samples/Linked List - Harsh Duche/PYTHON/SinglyLL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None

    # Create operation: Adding a new node to the end
    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    # Read operation: Accessing elements by index
    def get_node(self, index):
        current = self.head
        for i in range(index):
            if current is None:
                return -1  # Index out of bounds
            current = current.next
        return current.data if current else -1  # Node not found

    # Update operation: Modifying an element by index
    def update_node(self, index, data):
        current = self.head
        for i in range(index):
            if current is None:
                return
            current = current.next
        if current:
            current.data = data

    # Delete operation: Removing a node by index (without additional data structures)
    def delete_node(self, index):
        if index == 0:
            self.head = self.head.next
        else:
            current = self.head
            for i in range(index - 1):
                if current is None:
                    return
                current = current.next
            if current and current.next:
                current.next = current.next.next

    # isEmpty operation
    def is_empty(self):
        return self.head is None

# Example usage
linked_list = SinglyLinkedList()

# Create operation
linked_list.add_node(1)
linked_list.add_node(2)
linked_list.add_node(3)

# Read operation
print("Element at index 1:", linked_list.get_node(1))

# Update operation
linked_list.update_node(1, 4)

# Delete operation
linked_list.delete_node(1)

# isEmpty operation
print("Is the list empty?", linked_list.is_empty())

File: Code Samples/Linked List - Harsh Duche/CPP/SinglyLL.cpp
#include <iostream>

class Node {
public:
    int data;
    Node* next;

    Node(int data) : data(data), next(nullptr) {}
};

class SinglyLinkedList {
public:
    Node* head;

    SinglyLinkedList() : head(nullptr) {}

    // Create operation: Adding a new node to the end
    void addNode(int data) {
        Node* newNode = new Node(data);
        if (head == nullptr) {
            head = newNode;
        } else {
            Node* current = head;
            while (current->next != nullptr) {
                current = current->next;
            }
            current->next = newNode;
        }
    }

    // Read operation: Accessing elements by index
    int getNode(int index) {
        Node* current = head;
        for (int i = 0; i < index && current != nullptr; i++) {
            current = current->next;
        }
        return (current != nullptr) ? current->data : -1; // Index out of bounds
    }

    // Update operation: Modifying an element by index
    void updateNode(int index, int data) {
        Node* current = head;
        for (int i = 0; i < index && current != nullptr; i++) {
            current = current->next;
        }
        if (current != nullptr) {
            current->data = data;
        }
    }

    // Delete operation: Removing a node by index (without additional data structures)
    void deleteNode(int index) {
        if (index == 0) {
            Node* temp = head;
            head = head->next;
            delete temp;
        } else {
            Node* current = head;
            for (int i = 0; i < index - 1 && current != nullptr; i++) {
                current = current->next;
            }
            if (current != nullptr && current->next != nullptr) {
                Node* temp = current->next;
                current->next = current->next->next;
                delete temp;
            }
        }
    }

    // isEmpty operation
    bool isEmpty() {
        return head == nullptr;
    }
};

